src/Adapter/Product/Image/QueryHandler/GetProductImagesHandler.php

```diff
@@ -69,13 +72,20 @@ public function __construct(
      */
     public function handle(GetProductImages $query): array
     {
-        //todo throw exception if not single shop
-        $images = $this->productImageRepository->getImages($query->getProductId(), $query->getShopConstraint());
+        if (!$query->getShopConstraint()->getShopId()) {
+            throw new InvalidShopConstraintException('Only single shop constraint is supported');
+        }
+
+        // we still use hardcoded AllShops constraint here to get images for all the shops
+        // but when we format the image we will check if it is cover for the shopId from query,
+        // because that is the only property of image that differs between shops
+        $images = $this->productImageRepository->getImages($query->getProductId(), ShopConstraint::allShops());
```

Comment:
```suggestion
        $images = $this->productImageRepository->getImages($query->getProductId(), ShopConstraint::allShops());
        $coverId = $this->productImageRepository->getCoverImageId($query->getProductId(), $query->getShopConstraint()->getShopId());
```

Reply:
The cover ID should be fetched outside of the loop to avoid multiple duplicate queries